transformer_lens/HookedSAE.py

```diff
@@ -0,0 +1,111 @@
+from typing import Dict, Union
+
+import einops
+import torch
+import torch.nn.functional as F
+from jaxtyping import Float
+from torch import nn
+
+from transformer_lens.hook_points import (  # Hooking utilities
+    HookedRootModule,
+    HookPoint,
+)
+from transformer_lens.HookedSAEConfig import HookedSAEConfig
+
+
+class HookedSAE(HookedRootModule):
+    """Hooked SAE.
+
+    Implements a standard SAE with a TransformerLens hooks for SAE activations
+
+    Designed for inference / analysis, not training. For training, see Joseph Bloom's SAELens (https://github.com/jbloomAus/SAELens)
+    """
+
+    def __init__(self, cfg: Union[HookedSAEConfig, Dict]):
+        super().__init__()
+        if isinstance(cfg, Dict):
+            cfg = HookedSAEConfig(**cfg)
+        elif isinstance(cfg, str):
+            raise ValueError("Please pass in a config dictionary or HookedSAEConfig object.")
+        self.cfg = cfg
+
+        self.W_enc = nn.Parameter(
+            torch.nn.init.kaiming_uniform_(
+                torch.empty(self.cfg.d_in, self.cfg.d_sae, dtype=self.cfg.dtype)
+            )
+        )
+        self.W_dec = nn.Parameter(
+            torch.nn.init.kaiming_uniform_(
+                torch.empty(self.cfg.d_sae, self.cfg.d_in, dtype=self.cfg.dtype)
+            )
+        )
+        self.b_enc = nn.Parameter(torch.zeros(self.cfg.d_sae, dtype=self.cfg.dtype))
+        self.b_dec = nn.Parameter(torch.zeros(self.cfg.d_in, dtype=self.cfg.dtype))
+
+        self.hook_sae_input = HookPoint()
+        self.hook_sae_acts_pre = HookPoint()
+        self.hook_sae_acts_post = HookPoint()
+        self.hook_sae_recons = HookPoint()
+        self.hook_sae_error = HookPoint()
+        self.hook_sae_output = HookPoint()
+
+        self.to(self.cfg.device)
+
```

Comment:
You need a self.setup(), otherwise sae.run_with_cache and sae.run_with_hooks don't work. It would also be good to have a test to verify that you can do that, since clearly this is a gap in the tests!